### ISSUE TITLE
Container Exits with code 1

### ISSUE THREAD

**Issue by mcclede** [bitnami/bitnami-docker-magento#213]
**Description**

After running the Magento image for a while, if the container is shut down (e.g., docker compose down), it will not restart. It exits with code 1 every time it is restarted.

**Steps to reproduce the issue:**

1. Bring up Magento with the attached Docker Compose file
2. Let it run for awhile. I'm not sure how long, but it seems to work fine for a few days (2-3)
3. Bring down the configuration
4. Bring up the configuration.

**Describe the results you received:**

Everything appears to startup fine. However, after about 10 seconds the Magento container emits "magento_1 exited with code 1"

**Describe the results you expected:**

I expected everything to start and stay running with no errors

**Additional information you deem important (e.g. issue happens only occasionally):**

If I restart all images from scratch using the Docker Compose file everything works great for a few days, but then won't restart. I've done this serval times. How do I determine what is causing the exit "with code 1"?

Docker Compose file with necessary SQL script:
[Magento.zip](https://github.com/bitnami/bitnami-docker-magento/files/7341532/Magento.zip)

Start Up Logs:
[Startup Log.txt](https://github.com/bitnami/bitnami-docker-magento/files/7341548/Startup.Log.txt)

**Version**

- Output of `docker version`:

```
PS D:\CM> docker version
Client:
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:58:50 2021
 OS/Arch:           windows/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:10 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

- Output of `docker info`:

```
PS D:\CM> docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Build with BuildKit (Docker Inc., v0.6.1-docker)
  compose: Docker Compose (Docker Inc., v2.0.0-rc.3)
  scan: Docker Scan (Docker Inc., v0.8.0)

Server:
 Containers: 3
  Running: 2
  Paused: 0
  Stopped: 1
 Images: 5
 Server Version: 20.10.8
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: e25210fe30a0a703442421b0f60afac609f950a3
 runc version: v1.0.1-0-g4144b63
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 5.4.72-microsoft-standard-WSL2
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 25GiB
 Name: docker-desktop
 ID: SQ66:SYED:2U4B:FROC:SOXF:BQUP:LOVA:RXGZ:2WNY:UW2O:TLRQ:IS3U
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No blkio throttle.read_bps_device support
WARNING: No blkio throttle.write_bps_device support
WARNING: No blkio throttle.read_iops_device support
WARNING: No blkio throttle.write_iops_device support
```

- Output of `docker-compose version` (if applicable):

```
PS D:\CM> docker compose version
Docker Compose version v2.0.0-rc.3
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**

Device name	ANGEL3
Processor	AMD Ryzen 5 5600X 6-Core Processor                3.70 GHz
Installed RAM	32.0 GB
Device ID	72A3BAB3-9907-4ECE-8E5D-A027FD62088B
Product ID	00330-80000-00000-AA017
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

Edition	Windows 10 Pro
Version	20H2
Installed on	‎5/‎5/‎2021
OS build	19042.1237
Experience	Windows Feature Experience Pack 120.2212.3530.0

**Comment by randradas** [comment]
hello @mcclede thank you for such detailed issue, could you please reproduce the problem passing the following env variable to the container:

BITNAMI_DEBUG=true

It should help to be able to see better debugging output.

**Comment by CyanoFresh** [comment]
Happened to me 2 times. The last one after about 2 days after clean install.  

Logs of container before exit with code 1:
```
magento 17:50:45.31 

magento 17:50:45.31 Welcome to the Bitnami magento container

magento 17:50:45.31 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-magento

magento 17:50:45.31 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-magento/issues

magento 17:50:45.31 

magento 17:50:45.32 INFO  ==> ** Starting Magento setup **

magento 17:50:45.36 DEBUG ==> Validating settings in APACHE_* environment variables

magento 17:50:45.44 INFO  ==> Configuring PHP options

1

Cache types config flushed successfully

Cache cleared successfully

File system cleanup:

/bitnami/magento/generated/code/Composer

/bitnami/magento/generated/code/Magento

/bitnami/magento/generated/code/Psr

/bitnami/magento/generated/code/Symfony

The directory '/bitnami/magento/generated/metadata/' doesn't exist - skipping cleanup

Updating modules:

The "config.php" deployment config file isn't writable.

magento 17:50:45.44 INFO  ==> Setting PHP opcache.enable option

magento 17:50:45.45 INFO  ==> Setting PHP memory_limit option

magento 17:50:45.49 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars

magento 17:50:45.54 DEBUG ==> Validating settings in MAGENTO_* environment variables...

magento 17:50:45.55 WARN  ==> You set the environment variable ALLOW_EMPTY_PASSWORD=yes. For safety reasons, do not use this flag in a production environment.

magento 17:50:45.62 INFO  ==> Restoring persisted Magento installation

magento 17:50:55.59 INFO  ==> Trying to connect to the database server

magento 17:50:55.60 DEBUG ==> Getting configuration path '/db/connection/default/dbname' from Magento configuration

magento 17:50:55.63 DEBUG ==> Getting configuration path '/db/connection/default/username' from Magento configuration

magento 17:50:55.66 DEBUG ==> Getting configuration path '/db/connection/default/password' from Magento configuration

magento 17:50:55.70 DEBUG ==> Getting configuration path '/db/connection/default/host' from Magento configuration

magento 17:50:55.73 DEBUG ==> Executing SQL command:

SELECT 1

magento 17:50:55.74 INFO  ==> Trying to connect to Elasticsearch

timeout reached before the port went into state "inuse"

magento 17:51:05.77 INFO  ==> Upgrading database schema
```

docker version:

```
alex@Alex-PC:~/magento/magento_data$ docker version
Client: Docker Engine - Community
 Cloud integration: 1.0.17
 Version:           20.10.8
 API version:       1.41
 Go version:        go1.16.6
 Git commit:        3967b7d
 Built:             Fri Jul 30 19:54:02 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.8
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.16.6
  Git commit:       75249d8
  Built:            Fri Jul 30 19:52:10 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.9
  GitCommit:        e25210fe30a0a703442421b0f60afac609f950a3
 runc:
  Version:          1.0.1
  GitCommit:        v1.0.1-0-g4144b63
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Comment by marcosbc** [comment]
Hi @CyanoFresh, note that Magento starts as user `root` by default. If you initialize as `root` but later on restart the container as `1001`, it shouldn't work. The container must be run as the same user that initialized it.

Could it be that you changed the configuration after the first deployment, causing the user to change?

**Comment by CyanoFresh** [comment]
yes, the change of the user after first deployment fixes the topic starter's problem

**Comment by mcclede** [comment]
I was finally able to recreate the kind of problem I was having. The context is that I am attempting to develop a custom extension containing only a web API. During that process I had to take down my containers before I could ensure that everything was "working". Now when I attempt to restart the containers, using BITNAMI_DEBUG, I get the log below. From what I've been able to determine, I need to run setup:di:compile to fix the problem. I tried adding:

command: cd /bitnami/magento && php bin/magento setup:di:compile

to the docker-compose.yaml file, but that just caused other problems. So how can I run the setup:di:compile command, and possibly other commands, before a problem occurs with starting Magento?

magento  | magento 03:59:16.83
magento  | magento 03:59:16.83 Welcome to the Bitnami magento container
magento  | magento 03:59:16.83 Subscribe to project updates by watching https://github.com/bitnami/bitnami-docker-magento
magento  | magento 03:59:16.83 Submit issues and feature requests at https://github.com/bitnami/bitnami-docker-magento/issues
magento  | magento 03:59:16.83
magento  | magento 03:59:16.83 INFO  ==> ** Starting Magento setup **
magento  | magento 03:59:16.84 DEBUG ==> Validating settings in APACHE_* environment variables
magento  | magento 03:59:16.85 INFO  ==> Configuring PHP options
magento  | magento 03:59:16.86 INFO  ==> Validating settings in MYSQL_CLIENT_* env vars
magento  | magento 03:59:16.88 DEBUG ==> Validating settings in MAGENTO_* environment variables...
magento  | magento 03:59:16.90 INFO  ==> Restoring persisted Magento installation
magento  | magento 03:59:21.87 INFO  ==> Trying to connect to the database server
magento  | magento 03:59:21.89 DEBUG ==> Getting configuration path '/db/connection/default/dbname' from Magento configuration
magento  | magento 03:59:21.91 DEBUG ==> Getting configuration path '/db/connection/default/username' from Magento configuration
magento  | magento 03:59:21.93 DEBUG ==> Getting configuration path '/db/connection/default/password' from Magento configuration
magento  | magento 03:59:21.95 DEBUG ==> Getting configuration path '/db/connection/default/host' from Magento configuration
magento  | magento 03:59:21.96 DEBUG ==> Executing SQL command:
magento  | SELECT 1
magento  | 1
magento  | magento 03:59:21.97 INFO  ==> Trying to connect to Elasticsearch
magento  | magento 03:59:24.98 INFO  ==> Upgrading database schema
magento  | Cache types config flushed successfully
magento  | Class Magento\PageCache\Model\Config\Interceptor does not exist

**Comment by mcclede** [comment]
I resolved the issue I was having by adding:
```console
entrypoint: /bin/bash -c 'cd /bitnami/magento; php bin/magento setup:di:compile'
```
to my docker-compose.yaml file and bringing up the composition. When the setup:di:compile completed, the magento container exited with status 0. I then brought down the composition, removed the "entrypoint". I was then able to bring up the composition correctly. This doesn't seem like the "right" way to me. Seems like there should be a place to mount a volume containing scripts to be executed before and after Magento starts.

**Comment by mcclede** [comment]
That wasn't quite a perfect solution. It left the generated files owned by root instead of daemon. Once I got the container up, I had to go back and delete the "generated" folder and re-issue the setup:di:compile command. I probably should have added 'su daemon -s /bin/bash' to the lines executed by the entrypoint.

**Comment by MagnusSWE** [comment]
Just thought I should share how I solved a similar issue. My Magento container exited with code 255 during startup when restarting it after it ran nice for a while. Turned out I had used another user in the container so generated directory couldn't be emptied/removed during start process. I removed the volume magento-bitnami_magento_data with the faulty file owner and it could start just fine again. 

To see the volumes:
docker volume ls

To bring down the containers:
docker-compose down

To remove the volume with the failing file permissions:
docker volume rm magento-bitnami_magento_data
